File: algorithms/2_medium/L/03738_longest_non_decreasing_subarray_after_replacing_at_most_one_element/prefix_suffix.c
static inline int max_int(const int a, const int b) {
    return a > b ? a : b;
}

struct cell {
    int prefix;
    int suffix;
};

int longestSubarray(const int* nums, const int len) {
    assert(len >= 1 && len <= 100000 && nums != NULL);
    if (len <= 2) { return len; }

    struct cell* buff = malloc((size_t)len * sizeof(*buff));
    assert(buff != NULL);

    int prefix_cnt = 0;
    int prev_val = 0;
    for (int i = 0; i < len; i += 1) {
        const int num = nums[i];
        if (prev_val <= num) {
            prefix_cnt += 1;
        } else {
            prefix_cnt = 1;
        }
        prev_val = num;
        buff[i].prefix = prefix_cnt;
    }

    int suffix_cnt = 0;
    int next_val = 0;
    for (int i = len - 1; i >= 0; i -= 1) {
        const int num = nums[i];
        if (num <= next_val) {
            suffix_cnt += 1;
        } else {
            suffix_cnt = 1;
        }
        next_val = num;
        buff[i].suffix = suffix_cnt;
    }

    int res = 0;
    for (int i = 1; i + 1 < len; i += 1) {
        if (nums[i - 1] <= nums[i + 1]) {
            const int pref = buff[i - 1].prefix;
            const int suff = buff[i + 1].suffix;
            res = max_int(res, 1 + pref + suff);
        } else {
            res = max_int(res, max_int(1 + buff[i].suffix, 1 + buff[i].prefix));
        }
    }

/* cleanup: */
    free(buff);
    return res;
}

File: algorithms/2_medium/L/03738_longest_non_decreasing_subarray_after_replacing_at_most_one_element/prefix_suffix.rs
impl Solution {
    pub fn longest_subarray(nums: Vec<i32>) -> i32 {
        let len = nums.len();
        assert!(len <= 100_000);
        if len <= 2 { return len as i32; }

        let mut buff: Vec<(i32, i32)> = vec![(0, 0); len];

        let mut prefix_cnt = 0_i32;
        let mut prev_val = 0_i32;
        for i in 0..len {
            let num = nums[i];
            if prev_val <= num {
                prefix_cnt += 1;
            } else {
                prefix_cnt = 1;
            }
            prev_val = num;
            buff[i].0 = prefix_cnt;
        }

        let mut suffix_cnt = 0_i32;
        let mut next_val = 0_i32;
        for i in (0..len).rev() {
            let num = nums[i];
            if num <= next_val {
                suffix_cnt += 1;
            } else {
                suffix_cnt = 1;
            }
            next_val = num;
            buff[i].1 = suffix_cnt;
        }

        let mut res = 0_i32;
        for i in 1..(len - 1) {
            if nums[i - 1] <= nums[i + 1] {
                let pref = buff[i - 1].0;
                let suff = buff[i + 1].1;
                res = res.max(1 + pref + suff);
            } else {
                res = res.max(1 + buff[i].0).max(1 + buff[i].1);
            }
        }

        res
    }
}
